### PR TITLE
test(configgrpc|confighttp): sample test showcasing issue 10093

### PR DESCRIPTION
**Pull request by grandwizard28**
#### Description
This is a sample test showcasing issue https://github.com/open-telemetry/opentelemetry-collector/issues/10093

### CODE REVIEW

**Comment by grandwizard28** [comment]
@jpkrohling same test showcasing issue https://github.com/open-telemetry/opentelemetry-collector/issues/10093

File: config/configgrpc/configgrpc_test.go [hunk 1]
```diff
@@ -963,6 +963,118 @@ func TestClientInfoInterceptors(t *testing.T) {
...
+					IncludeMetadata: tC.includeMetadata,
+				}
+				gss.Auth = &configauth.Authentication{
+					AuthenticatorID: mockID,
+				}
+				host := &mockHost{
+					ext: map[component.ID]component.Component{
+						mockID: auth.NewServer(auth.WithServerAuthenticate(func(ctx context.Context, headers map[string][]string) (context.Context, error) {
+							return client.NewContext(ctx, client.Info{
+								Metadata: client.NewMetadata(metadata.Pairs("some-key-set-in-auth", "some-value-set-in-auth")),
```
Comment (jpkrohling) [inline config/configgrpc/configgrpc_test.go:1012]: I see what you mean now. Authenticators are supposed to only provide data via the AuthData, not Metadata. Metadata is only supposed to be set by the parts that deal with the connection with the client, such as gRPC and HTTP servers (like those in receivers). Your downstream consumer of this information would have to call AuthData.GetAttribute(...). 

https://github.com/open-telemetry/opentelemetry-collector/blob/3d0189df8a26a5d654b51080dcd07ae57d394091/client/client.go#L26-L30

Here's an implementation example:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5cc9e03d054728b8f97d9214aa58910aa493c502/extension/basicauthextension/extension.go#L105-L107

Comment (grandwizard28) [reply config/configgrpc/configgrpc_test.go:1012]: Yeah, but consider the following:

- This doesnt happen for oltp/http
- If I try putting in the AuthData, the batch processor will drop it. See this: https://github.com/open-telemetry/opentelemetry-collector/pull/10002

I feel the behaviours should be similar for oltp/http and oltp/grpc. Wouldn't you agree?

Comment (jpkrohling) [reply config/configgrpc/configgrpc_test.go:1012]: Can you add a test case here for otlp/http? There should be no difference between the receivers, and if the authenticator is overriding the metadata, none should see the original headers after the authenticator runs.

Comment (grandwizard28) [reply config/configgrpc/configgrpc_test.go:1012]: Table depicting what happens to metadata set in Auth:

| Protocol/Config | Include Metadata = True | Include Metadata = False |
| ------ | ------ | ----- |
| GRPC | overriden | not overriden |
| HTTP | not overriden | not overriden |

Comment (grandwizard28) [reply config/configgrpc/configgrpc_test.go:1012]: @jpkrohling does the above make sense?

File: config/configgrpc/configgrpc_test.go [hunk 2]
```diff
@@ -963,6 +963,118 @@ func TestClientInfoInterceptors(t *testing.T) {
...
+					},
+					IncludeMetadata: tC.includeMetadata,
+				}
+				gss.Auth = &configauth.Authentication{
+					AuthenticatorID: mockID,
+				}
+				host := &mockHost{
+					ext: map[component.ID]component.Component{
+						mockID: auth.NewServer(auth.WithServerAuthenticate(func(ctx context.Context, headers map[string][]string) (context.Context, error) {
+							return client.NewContext(ctx, client.Info{
```
Comment (jpkrohling) [inline config/configgrpc/configgrpc_test.go:1011]: The right thing here would be to obtain the client from the context, set _only_ the Auth attributes, and wrap it in the context that is being returned, like this:

```go
mockID: auth.NewServer(auth.WithServerAuthenticate(func(ctx context.Context, headers map[string][]string) (context.Context, error) {
	cl := client.FromContext(ctx)
	cl.Auth = &mockAuthData{
		Attributes: map[string]string{"some-key-set-in-auth": "some-value-set-in-auth"},
	}

	return client.NewContext(ctx, cl), nil
})),
```

File: config/configgrpc/configgrpc_test.go [hunk 3]
```diff
@@ -963,6 +963,118 @@ func TestClientInfoInterceptors(t *testing.T) {
...
+				grpcClientConn, errClient := gcs.ToClientConn(context.Background(), componenttest.NewNopHost(), tt.TelemetrySettings())
+				require.NoError(t, errClient)
+				defer func() { assert.NoError(t, grpcClientConn.Close()) }()
+
+				cl := ptraceotlp.NewGRPCClient(grpcClientConn)
+				ctx, cancelFunc := context.WithTimeout(context.Background(), 2*time.Second)
+				defer cancelFunc()
+
+				// test
+				tC.tester(ctx, cl)
```
Comment (jpkrohling) [inline config/configgrpc/configgrpc_test.go:1058]: This is counter-intuitive, as we set the headers on the ClientSettings, but the client making use of the construct is responsible for using the headers. So, the headers you are setting at ClientConfig aren't being sent to the actual service. You'll need something like this:

```go
// this is what we expect clients to do before making a gRPC call
headers := map[string]string{}
for k, v := range gcs.Headers {
	headers[k] = string(v)
}
md := metadata.New(headers)
ctx = metadata.NewOutgoingContext(ctx, md)
```

File: config/confighttp/confighttp_test.go [hunk 4]
```diff
@@ -1207,6 +1207,83 @@ func TestServerAuth(t *testing.T) {
...
+			desc:            "metadata:false",
+			includeMetadata: false,
+		},
+	}
+
+	for _, tC := range testCases {
+		t.Run(tC.desc, func(t *testing.T) {
+			host := &mockHost{
+				ext: map[component.ID]component.Component{
+					mockID: auth.NewServer(auth.WithServerAuthenticate(func(ctx context.Context, headers map[string][]string) (context.Context, error) {
```
Comment (jpkrohling) [inline config/confighttp/confighttp_test.go:1229]: Same comments as gRPC.

**Review by jpkrohling — COMMENTED** [review]
I realize that perhaps the API isn't very intuitive at the moment, and we either need better documentation or to prevent the API from being misused. 

I found a couple of inconsistencies with the test, and I'm sending a new commit to this PR to show how we expect the APIs to be used.

**Comment by grandwizard28** [comment]
Right! The intended usage is to set only the AuthData in custom authenticators and not touch metadata at all.

If someone tries to set the metadata in a custom authenticator (as I did), http would allow it to propagate but grpc wont. (With IncludeMetadata: true) Even though this is not the intended usage, the two protocols behave differently in this instance.

My Reason for doing so:
The whole reason I tried to set Metadata in the Authenticator was because the batch processor was dropping the AuthData altogether.

All I want is to be able to propagate Authdata across the collector.

**Comment by jpkrohling** [comment]
> All I want is to be able to propagate Authdata across the collector

You can do that today already, except if you use the batching processor: that one will group data from potentially different connections into one single batch. Those connections came with potentially different auth data.

The proper solution would be to enhance the batch processor, so that it can also group data by auth data in addition to metadata.

**Comment by jpkrohling** [comment]
@grandwizard28 , do you think this should be closed?

**Pull request closed without merging**